[Docs] Update torch.(squeeze, split, set_printoptions, save) docs.

KushajveerSingh opened the pull request:
I added the following to the docs:
1. `torch.save`. 
    1. Added doc for `_use_new_zipfile_serialization` argument.
    2. Added a note telling that extension does not matter while saving.
    3. Added an example showing the use of above argument along with `pickle_protocol=5`.

2. `torch.split`
    1. Added an example showing the use of the function.

3. `torch.squeeze`
   1. Added a warning for batch_size=1 case.

4. `torch.set_printoptions`
    1. Changed the docs of `sci_mode` argument from
        ```
        sci_mode: Enable (True) or disable (False) scientific notation. If
                 None (default) is specified, the value is defined by `_Formatter`
        ```
        to
        ```
        sci_mode: Enable (True) or disable (False) scientific notation. If
                 None (default=False) is specified, the value is defined by 
                `torch._tensor_str._Formatter`.
        ```

KushajveerSingh committed:
update docs of torch.save

Add doc for _use_new_zipfile_serialization argument, add a note telling that any file extension can be used with torch.save. And add an example showing the use of above argument with pickle_protocol=5.

KushajveerSingh committed:
Update torch.squeeze docs

Added a warning in torch.squeeze when using batch_size=1.

KushajveerSingh committed:
Update torch.split docs

Added an example showing the usage of torch.split

KushajveerSingh committed:
Update docs of sci_mode in set_printoptions

KushajveerSingh commented:
I made the changes to the following files to update the docs.
1. `torch.save` -> `torch.serialization.py`
2. `torch.squeeze` -> `torch._torch_docs.py`
3. `torch.split` -> `torch.functional.py`
4. `torch.set_printoptions` -> `torch._tensor_str.py`

vadimkantorov commented on torch/serialization.py:
```diff
@@ -338,7 +338,13 @@ def save(obj, f, pickle_module=pickle, pickle_protocol=DEFAULT_PROTOCOL, _use_ne
+    
+    Note:
+        The file extension does not matter while saving the object. The object can be saved using
+        .pt, .pth, .zip or any other extension. You can literally save the object as .py file and 
```
If possible, I propose to avoid recommending/suggesting `*.pth` as reasonable extension for serialization, because "pth" is a special file extension for Python and can lead no nasty package importing errors: https://github.com/pytorch/pytorch/issues/14864

zou3519 commented on torch/_tensor_str.py:
```diff
@@ -38,7 +38,8 @@ def set_printoptions(
             the above options. (any one of `default`, `short`, `full`)
         sci_mode: Enable (True) or disable (False) scientific notation. If
-            None (default) is specified, the value is defined by `_Formatter`
+            None (default=False) is specified, the value is defined by 
```
`None` is the default and means the framework makes a decision on whether or not to print with scientific notation. See https://github.com/pytorch/pytorch/blob/caaea084e9e63dd3f22edcab0fda6822f7634820/torch/_tensor_str.py#L117-L142 . We should keep this text as `None (default)` but make it more clear that the framework chooses when it happens.

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -5510,6 +5510,9 @@ def merge_dicts(*dicts):
 .. note:: The returned tensor shares the storage with the input tensor,
           so changing the contents of one will change the contents of the other.
 
+.. warning:: If batch_size=1, then `squeeze(input)` will also remove the batch
```
Not all tensors have something called a "batch_size". Could we write something like "if the tensor has a batch dimension of size 1" ?

zou3519 commented on torch/serialization.py:
```diff
@@ -338,7 +338,13 @@ def save(obj, f, pickle_module=pickle, pickle_protocol=DEFAULT_PROTOCOL, _use_ne
         pickle_module: module used for pickling metadata and objects
         pickle_protocol: can be specified to override the default protocol
-
+        _use_new_zipfile_serialization: saves the object as a zip file rather than a binary file
```
This flag looks like it should not be exposed to the public because of the leading underscore.

@eellison If I remember correctly, this is a TorchScript thing. Do you know what this flag is and if we should document it?

eellison replied:
hey don't really know about it, maybe @zdevito or @voznesenskym would know

voznesenskym replied:
Hey! qq, when you say "to the public", do you mean as part of the API? Or do you mean outside of fbcode?

zou3519 replied:
I mean as a part of the API. Generally as a "pytorch user", when I see something with underscores prepended to it, that means that it is not part of the public API

voznesenskym replied:
As I understand it, the purpose of this is for a migration. Having it as part of the API for now seems like a relatively simple way to denote which serialization you want to use. I am super open to hearing alternative ways of implementing the same functionality though!

KushajveerSingh committed:
remove .pth note from torch.save docs

Added a new note, telling that .pt extension is used in most cases. And removed .pth extension from the next note. Also, wrapped the last line of the note in brackets, where it is said that pytorch objects can be saved as .py files.

KushajveerSingh committed:
correct spelling typo intented->intended in save

KushajveerSingh committed:
update set_printoptions docs

Changed default=False to default in sci_mode, as the value of sci_mode is framework dependent and is chosen automatically. Also, added a line to the docs of sci_mode making it more clear that this value is chosen automatically.

KushajveerSingh committed:
update torch.squeeze warning docs

KushajveerSingh commented:
@vadimkantorov Thanks for that suggestion. I didn't know `.pth` was also used in python also. I changed the docs as follows for `torch.save`

```
Note:
    `.pt` extension is used for most of the cases.
    
Note:
    The file extension does not matter while saving the object. The object can be saved using
    .pt, .zip or any other extension. (You can literally save the object as .py file and 
    it will just work as intended).
```

@zou3519 Thanks for the pointer, I missed `self.init_mode`. I changed the docs for sci_mode as follows

```
sci_mode: Enable (True) or disable (False) scientific notation. If
        None (default) is specified, the value is defined by 
        `torch._tensor_str._Formatter`. This value is automatically chosen
        by the framework.
```

It makes sense to keep the value to default rather than default=False. Also, I added a line telling that this value is adjusted automatically. 

@zou3519 I changed the wording of `torch.squeeze` warning.

```
.. warning:: If the tensor has a batch dimension of size 1, then `squeeze(input)` 
          will also remove the batch dimension, which can lead to unexpected 
          errors.
```

zou3519 commented on torch/serialization.py:
```diff
@@ -338,7 +338,16 @@ def save(obj, f, pickle_module=pickle, pickle_protocol=DEFAULT_PROTOCOL, _use_ne
-
+        _use_new_zipfile_serialization: saves the object as a zip file rather than a binary file
+    
+    Note:
+        `.pt` extension is used for most of the cases.
+    
+    Note:
+        The file extension does not matter while saving the object. The object can be saved using
+        .pt, .zip or any other extension. (You can literally save the object as .py file and 
+        it will just work as intended). 
```
Let's combine these two as 
```
Note:
    A common PyTorch convention is to save tensors using either a .pt or .pth file extension.
```

I would leave out the "The object can be saved using
        .pt, .zip or any other extension. (You can literally save the object as .py file and 
        it will just work as intended). " part, we don't want users to save their files as .zip or .py and then attempt to run `winzip` or `python` on them.

vadimkantorov replied:
@zou3519 Recommending `*.pth` is not very nice, because `*.pth`, because "pth" is a special file extension for Python and can lead no nasty package importing errors: #14864 ...

zou3519 replied:
Sounds good, let's avoid recommending `*.pth` then. We should change our recommendation in the tutorials then: https://pytorch.org/tutorials/beginner/saving_loading_models.html, but that can be done separately.

KushajveerSingh replied:
Done

KushajveerSingh committed:
udpate torch.save docs

Removed the documentation for _use_new_zipfile_serialization as it might be an internal torch function. Also, updated the note to tell that .pt is a common pytorch convention for saving tensors. Removed the example showing the usage of _use_new_zipfile_serialization.

KushajveerSingh commented on torch/serialization.py:
```diff
@@ -338,7 +338,16 @@ def save(obj, f, pickle_module=pickle, pickle_protocol=DEFAULT_PROTOCOL, _use_ne
         pickle_module: module used for pickling metadata and objects
         pickle_protocol: can be specified to override the default protocol
-
+        _use_new_zipfile_serialization: saves the object as a zip file rather than a binary file
```
Done

voznesenskym replied:
What this does is it changes the serialization implementation. Old and new are identically compatible. However, moving the serialization into our ownership is the first step towards python/C++ compatible saves+loads. You probably dont need to document this flag. cc: @suo for a definitive answer.

zou3519 reviewed (approved):
lgtm, thanks @KushajveerSingh. We can always follow up on how to document the `_use_new_zipfile_serialization` flag, so I'll merge this commit as-is.

zou3519 commented on torch/serialization.py:
```diff
@@ -338,7 +338,10 @@ def save(obj, f, pickle_module=pickle, pickle_protocol=DEFAULT_PROTOCOL, _use_ne
         pickle_module: module used for pickling metadata and objects
         pickle_protocol: can be specified to override the default protocol
-
+    
```
The Linter is complaining that lines 341 and lines 344 contains whitespace: https://github.com/pytorch/pytorch/pull/39303/checks?check_run_id=739499514

KushajveerSingh replied:
Resolved the issue and rebased the repo as well.

zou3519 reviewed (changes requested):
The linter is complaining: https://github.com/pytorch/pytorch/pull/39303/checks?check_run_id=739499514 .

Looking at the tests, you might have to rebase the PR as well (because our CI can only handle "recent" commits, where recent is approximately a few days old)

KushajveerSingh committed:
Merge branch 'master' of github.com:pytorch/pytorch into doc-changes

KushajveerSingh committed:
remove white spaces

KushajveerSingh committed:
remove white spaces
